lib/core/base/virtual-node.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * Wrap the real node and provide list of the flattened children
+ *
+ * @param node {Node} - the node in question
+ * @param shadowId {String} - the ID of the shadow DOM to which this node belongs
+ */
+function VirtualNode(node, shadowId) {
+	this.shadowId = shadowId;
+	this.children = [];
+	this.actualNode = node;
+
+	this._isHidden = null; // will be populated by axe.utils.isHidden
+	this._cache = {};
+
+	// abstract Node and Element APIs so we can run axe in DOM-less
+	// environments. these are static properties in the assumption
+	// that axe does not change any of them while it runs.
+	this.elementNodeType = node.nodeType;
+	this.elementNodeName = node.nodeName.toLowerCase();
+	this.elementId = node.id;
+
+	if (axe._cache.get('nodeMap')) {
+		axe._cache.get('nodeMap').set(node, this);
+	}
+}
+
+/**
+ * Determine if the actualNode has the given class name.
+ * @see https://j11y.io/jquery/#v=2.0.3&fn=jQuery.fn.hasClass
+ * @param {String} className - The class to check for.
+ * @return {Boolean} True if the actualNode has the given class, false otherwise.
+ */
+VirtualNode.prototype.hasClass = function(className) {
+	if (typeof this.actualNode.className !== 'undefined') {
```

Comment:
Please refactor this to keep the happy-path closer to the left.

```js
if (typeof this.actualNode.className === 'undefined') {
  return
}

const selector = ' ' + className +' '
return ' ' + this.actualNode.className + ' '.indexOf(selector) >= 0
```

Reply:
Also, it appears we expect `actualNode.className` to be of type `string`. This will cause nasal demons:

```js
this.actualNode.className = { hello: false }
this.hasClass('potato')
```

Reply:
See https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88 for a good article/primer about happy-path.

Reply:
`className` always returns a string, even if you set it to an object it will return `[object Object]` since it reflects the class attribute value.